Add interactive hosted notebook with examples.

desilinguist opened the pull request:
Originally created by @AVajpayeeJr via a fork. Moving here for easier editing of notebook.

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+{
```
Nitpick: Can we make the cells be sequential and starting at 1?

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As you can see, SKLL configuration files also support Python comments. In this particular example, we are going to do cross-validation on the IRIS training set with 4 different learners. The hyperparameters for each learner are turned by searching a [default grid of values](https://skll.readthedocs.io/en/latest/run_experiment.html#param-grids-optional) defined in SKLL (the default grid can b ...
```
Nitpick: I think it should be "Python-style comments".

mulhod replied:
Also: typo "turned" -> "tuned"

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+   "source": [
+    "### 1c. Running the experiment\n",
+    "\n",
+    "After creating the configuration file, we can now run the experiment using the `run_experiment` command. This scripts only takes a single required argument –– the path to the configuration file. However, it also [supports other optional arguments](https://skll.readthedocs.io/en/latest/run_experiment.html#using-run-experiment) for other use cases, e.g., running an experiment on a DRMAA-compat ...
```
Nitpick: "e.g.," should probably just be "e.g." (no comma)

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+   "metadata": {},
+   "source": [
+    "We see that the subdirectory `iris/output` has been created and that it contains several new files:\n",
+    "- a plain text `.results` file that contains human-readable output of the results for each learner in our list\n",
```
Nitpick: There's no comma after the first bullet point (whereas there is one for every other bullet point after it that needs it). Also, I think typically there would be an "and" on the line preceding the last bullet point after the comma.

However, personally, I think all the commas should be dropped (and no "and" added as mentioned above).

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This file has many useful columns but let's just look at small subset for now."
```
Nitpick: Missing "a" before "subset"

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "As mentioned earlier, SKLL provides several [utility scripts](https://skll.readthedocs.io/en/latest/utilities.html) that can be run from the command line. We examined the `skll_convert` earlier. Next, let's examine the `print_model_weights` script which can print out the model parameters (weights) for linear models on the command line which can be quite useful for debugging and interpretabil ...
```
Remove "the" from before "\`skll_convert\`" or add "utility" afterwards.

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+    "\n",
+    "While the command line tools are intended to be the primary method of using SKLL for normal users who want to run batched machine-learning experiments, the SKLL Python API can also be useful for building applications on top of SKLL and for other advanced use cases. \n",
+    "\n",
+    "The [*learner*](https://skll.readthedocs.io/en/latest/api/learner.html) and [*data*](https://skll.readthedocs.io/en/latest/api/data.html) modules can be used to broadly replicate the functionalities of command line tools. For example, the `skll.data.Reader` class can be used to programmatically read any SKLL-compatible feature file into a SKLL [*FeatureSet*](https://skll.readthedocs.io/en/la ...
```
If we can, I think we should make the font the same for things like `skll.data.Reader` and `FeatureSet` (even if the latter is a link, which I think should still be possible). I prefer consistency here.

desilinguist replied:
So you'd prefer that it wasn't visibly a link? We won't get the underline ...

mulhod replied:
I thought there might be something like this going on but I thought I would point it out anyway just in case it could be made uniform. I think it's fine, then.

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `Learner` class can be used to read in saved models from disk into memory and also to create new learners from scratch. First let's load in one of the models from our previous IRIS experiment and then use to generate predictions for our test set."
```
Add "it" after "use"

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, let's create a new learner, train and tune it on the training set using accuracy."
```
It might be better to add " as the grid search objective function to the end".

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The `Learner.train()` method also returns `grid_search_results` which is a dictionary containing useful intermediate results from the grid-search process, e.g., the value of the objective for each of the parameter values in the grid and for each of the 3 cross-validation folds."
```
"e.g.," -> "e.g."

mulhod commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+       "      <td>0.000240</td>\n",
+       "      <td>0.000004</td>\n",
+       "      <td>100</td>\n",
+       "      <td>{'C': 10
```
For some reason, I see the following warnings when running this cell:
```
/srv/conda/envs/notebook/lib/python3.7/site-packages/sklearn/utils/deprecation.py:125: FutureWarning: You are accessing a training score ('split0_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
/srv/conda/envs/notebook/lib/python3.7/site-packages/sklearn/utils/deprecation.py:125: FutureWarning: You are accessing a training score ('split1_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
/srv/conda/envs/notebook/lib/python3.7/site-packages/sklearn/utils/deprecation.py:125: FutureWarning: You are accessing a training score ('split2_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
/srv/conda/envs/notebook/lib/python3.7/site-packages/sklearn/utils/deprecation.py:125: FutureWarning: You are accessing a training score ('mean_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
/srv/conda/envs/notebook/lib/python3.7/site-packages/sklearn/utils/deprecation.py:125: FutureWarning: You are accessing a training score ('std_train_score'), which will not be available by default any more in 0.21. If you need training scores, please set return_train_score=True
  warnings.warn(*warn_args, **warn_kwargs)
```

If there's anything we can do about this, that would be great.

mulhod replied:
It's hard to comment on a raw ipynb file. The line I meant was `pd.DataFrame(grid_search_results)`, to be clear.

desilinguist replied:
Yeah, this would go away once we merge in the scikit-learn 0.21.3 branch since we would no longer generate these fields.

desilinguist commented on examples/Tutorial.ipynb:
```diff
@@ -0,0 +1,1521 @@
+    "\n",
+    "- Get the data in a [SKLL compatible format](https://skll.readthedocs.io/en/latest/run_experiment.html#file-formats).\n",
+    "- Create a SKLL [experiment configuration file](https://skll.readthedocs.io/en/latest/run_experiment.html#create-config) for cross-validation.\n",
+    "- Run the machine learning experiment using the [*run_experiment*](https://skll.readthedocs.io/en/latest/run_experiment.html) command.\n",
```
I italicize it when I need it to be a link and monospaced otherwise. Monospaced links do not get the underline so the reader doesn't know it's a link.

mulhod replied:
That's unfortunate. Anyway, sounds good to me, then.

desilinguist commented:
@mulhod I addressed all your comments. Can you please approve? Note that you won't be able to launch the binder anymore (as I updated the badge to point to `master` in anticipation of the merge). But you should be able to look at the notebook locally? Thanks!
